File: src/main/java/com/example/css/CssModulesIndexedStylesVarPsiReferenceContributor.java
package com.example.css;
import com.intellij.lang.javascript.psi.JSLiteralExpression;
import com.intellij.openapi.util.Ref;
import com.intellij.openapi.util.TextRange;
import com.intellij.psi.*;
import com.intellij.psi.css.*;
import com.intellij.psi.util.PsiTreeUtil;
import com.intellij.util.ProcessingContext;
import org.apache.commons.lang3.StringUtils;
import org.jetbrains.annotations.NotNull;

import java.util.Optional;


/**
 * Adds a PSI references from an indexed string literal on a styles object to its corresponding class name.
 * For example, the 'normal' in styles['normal'] will point to the '.normal {}' CSS class in a requirement'd stylesheet.
 * If not Founded  , will point the First CssClassSelector in stylesheet
 */
public class CssModulesIndexedStylesVarPsiReferenceContributor extends PsiReferenceContributor {
    private static final String DOT = ".";
    @Override
    public void registerReferenceProviders(@NotNull PsiReferenceRegistrar registrar) {
        registrar.registerReferenceProvider(QCssModulesUtil.STRING_PATTERN, new PsiReferenceProvider() {
            @NotNull
            @Override
            public PsiReference @NotNull [] getReferencesByElement(@NotNull PsiElement element, @NotNull ProcessingContext context) {
                final PsiElement cssClassNamesImportOrRequire = QCssModulesUtil.getCssClassNamesImportOrRequireDeclaration((JSLiteralExpression) element);
                if (cssClassNamesImportOrRequire != null) {
                    final String literalClass = DOT + StringUtils.stripStart(StringUtils.stripEnd(element.getText(), "\"'"), "\"'");
                    final Ref<StylesheetFile> referencedStyleSheet = new Ref<>();
                    final PsiElement psiElement = QCssModulesUtil.getCssClass(cssClassNamesImportOrRequire, literalClass, referencedStyleSheet);
                    if (psiElement != null) {
                        return new PsiReference[]{new PsiReferenceBase<PsiElement>(element) {
                            @Override
                            public @NotNull PsiElement resolve() {
                                return psiElement;
                            }

                            @NotNull
                            @Override
                            public Object @NotNull [] getVariants() {
                                return new Object[0];
                            }
                        }};
                    }
                }
                return new PsiReference[0];
            }
        });
    }

}

File: src/main/kotlin/com/peppa/css/psi/CssModulesIndexedStylesVarPsiReferenceContributor.kt
package com.peppa.css.psi

import com.intellij.lang.javascript.JSTokenTypes
import com.intellij.lang.javascript.psi.JSFile
import com.intellij.lang.javascript.psi.JSIndexedPropertyAccessExpression
import com.intellij.lang.javascript.psi.JSLiteralExpression
import com.intellij.patterns.PlatformPatterns
import com.intellij.psi.PsiElement
import com.intellij.psi.PsiReference
import com.intellij.psi.PsiReferenceContributor
import com.intellij.psi.PsiReferenceProvider
import com.intellij.psi.PsiReferenceRegistrar
import com.intellij.util.ProcessingContext
import com.peppa.css.completion.findReferenceStyleFile

class CssModulesIndexedStylesVarPsiReferenceContributor : PsiReferenceContributor() {

    override fun registerReferenceProviders(registrar: PsiReferenceRegistrar) {
        registrar.registerReferenceProvider(INDEXED_ACCESS_FILTER, IndexedReferenceProvider())
    }

    companion object {
        private val INDEXED_ACCESS_FILTER = PlatformPatterns
            .psiElement(JSLiteralExpression::class.java)
            .withParent(JSIndexedPropertyAccessExpression::class.java)
            .inFile(PlatformPatterns.psiFile(JSFile::class.java))
    }
}

private class IndexedReferenceProvider : PsiReferenceProvider() {

    override fun getReferencesByElement(element: PsiElement, context: ProcessingContext): Array<PsiReference> {
        val literal = element as? JSLiteralExpression ?: return PsiReference.EMPTY_ARRAY

        if (literal.node.firstChildNode?.elementType != JSTokenTypes.STRING_LITERAL) {
            return PsiReference.EMPTY_ARRAY
        }

        val name = literal.stringValue?.trim()?.takeIf { it.isNotBlank() } ?: return PsiReference.EMPTY_ARRAY
        val styleFile = findReferenceStyleFile(literal) ?: return PsiReference.EMPTY_ARRAY

        return arrayOf(CssModuleClassReference(literal, styleFile, name))
    }
}

fun isStyleIndex(element: JSLiteralExpression): Boolean = findReferenceStyleFile(element) != null